File: packages/shared-types/src/id-generation.ts
import type {
  AirportCode,
  FAFlightID,
  InternalFlightLegID,
  ISODateTimeString,
  ProvisionalFlightLegID,
} from "./domain-types.js";
import { hashStableParts } from "./stable-hash.js";

export interface ProvisionalFlightLegIDInput {
  ident: string;
  originCode: AirportCode;
  destinationCode: AirportCode;
  scheduledOut: ISODateTimeString;
}

export interface InternalFlightLegIDInput {
  faFlightId: FAFlightID;
  originCode: AirportCode;
  destinationCode: AirportCode;
  scheduledOut: ISODateTimeString;
  legIndex: number;
}

export function generateProvisionalFlightLegId(
  input: ProvisionalFlightLegIDInput,
): ProvisionalFlightLegID {
  return `sched_${hashStableParts([
    input.ident,
    input.originCode,
    input.destinationCode,
    input.scheduledOut,
  ])}`;
}

export function generateInternalFlightLegId(input: InternalFlightLegIDInput): InternalFlightLegID {
  return `iflg_${hashStableParts([
    input.faFlightId,
    input.originCode,
    input.destinationCode,
    input.scheduledOut,
    input.legIndex,
  ])}`;
}

File: services/internal/domain/id_generation.go
package domain

import (
	"fmt"
	"hash/fnv"
	"strconv"
)

type ProvisionalFlightLegIDInput struct {
	Ident           string
	OriginCode      AirportCode
	DestinationCode AirportCode
	ScheduledOut    ISODateTimeString
}

type InternalFlightLegIDInput struct {
	FAFlightID      FAFlightID
	OriginCode      AirportCode
	DestinationCode AirportCode
	ScheduledOut    ISODateTimeString
	LegIndex        int
}

func GenerateProvisionalFlightLegID(input ProvisionalFlightLegIDInput) ProvisionalFlightLegID {
	return ProvisionalFlightLegID("sched_" + hashFlightIDParts(
		input.Ident,
		string(input.OriginCode),
		string(input.DestinationCode),
		string(input.ScheduledOut),
	))
}

func GenerateInternalFlightLegID(input InternalFlightLegIDInput) InternalFlightLegID {
	return InternalFlightLegID("iflg_" + hashFlightIDParts(
		string(input.FAFlightID),
		string(input.OriginCode),
		string(input.DestinationCode),
		string(input.ScheduledOut),
		strconv.Itoa(input.LegIndex),
	))
}

func hashFlightIDParts(parts ...string) string {
	hasher := fnv.New64a()
	for index, part := range parts {
		if index > 0 {
			_, _ = hasher.Write([]byte{0x1f})
		}
		_, _ = hasher.Write([]byte(part))
	}

	return fmt.Sprintf("%016x", hasher.Sum64())[:12]
}
